cpp/src/arrow/compute/kernels/scalar_temporal.cc

```diff
@@ -187,16 +243,28 @@ struct DayOfYear {
 
 template <typename Duration>
 struct ISOYear {
+  explicit ISOYear(const time_zone* tz = nullptr) : tz_(tz) {}
+
   template <typename T, typename Arg0>
-  static T Call(KernelContext*, Arg0 arg, Status*) {
-    const auto t = floor<days>(sys_time<Duration>(Duration{arg}));
+  T Call(KernelContext*, Arg0 arg, Status*) const {
+    if (tz_ == nullptr) {
+      const auto t = floor<days>(sys_time<Duration>(Duration{arg}));
+      auto y = year_month_day{t + days{3}}.year();
+      auto start = sys_time<days>((y - years{1}) / dec / thu[last]) + (mon - thu);
+      if (t < start) {
+        --y;
+      }
+      return static_cast<T>(static_cast<int32_t>(y));
+    }
+    const auto t = floor<days>(tz_->to_local(sys_time<Duration>(Duration{arg})));
     auto y = year_month_day{t + days{3}}.year();
-    auto start = sys_time<days>((y - years{1}) / dec / thu[last]) + (mon - thu);
+    auto start = local_days((y - years{1}) / dec / thu[last]) + (mon - thu);
     if (t < start) {
```

Comment:
It looks like we would like to be able to write:
```c++
template <typename Duration>
struct ISOYearExtractor {
  template <typename T, typename TimePointConverter, typename DaysConverter>
  static T Call(Duration d, TimePointConverter&& convert_timepoint, DaysConverter&& convert_days) {
    const auto t = floor<days>(convert_timepoint(sys_time<Duration>(d));
    auto y = year_month_day{t + days{3}}.year();
    auto start = convert_days((y - years{1}) / dec / thu[last]) + (mon - thu);
    if (t < start) {
      --y;
    }
    return static_cast<T>(static_cast<int32_t>(y));
  }
};
```

and then something like:
```c++

template <typename Duration, template <typename Duration> ExtractorType>
struct ExtractorExec {
  using Extractor = ExtractorType<Duration>;

  explicit ExtractorExec(const time_zone* tz = nullptr) : tz_(tz) {}

  template <typename T, typename Arg0>
  T Call(KernelContext*, Arg0 arg, Status*) const {
    const Duration d{arg};
    if (tz_ == nullptr) {
      return Extractor::Call(arg,
          [](sys_time<Duration> t) { return t; },
          [](sys_days d) { return d; });
    }
    return Extractor::Call(arg,
        [this](sys_time<Duration> t) { return tz_->to_local(t); },
        [](sys_days d) { return local_days(d); });
    }
  };
};

template <typename Duration>
using ISOYear = ExtractorExec<Duration, ISOYearExtractor>;
// same for other extractors...
```

Reply:
I've adjusted your proposal to work with `ScalarUnaryNotNullStateful`.